### ISSUE TITLE
"Never here" message with realSFS fst index

### ISSUE THREAD

**Issue by TeresaPegan** [ANGSD/angsd#331]
Hello,
I've been using realSFS fst index to do some pairwise fst calculations, and I am sometimes getting output that says "Never here." As in,

```
 $ANGSDIR/misc/realSFS fst index $GUTTDIR/MB1910.saf.idx $GUTTDIR/MB1927.saf.idx $GUTTDIR/MB27B.saf.idx $GUTTDIR/zo11099.saf.idx -sfs $GUTTDIR/MB1910.MB1927.ml -sfs $GUTTDIR/MB1910.MB27B.ml -sfs $GUTTDIR/MB1910.zo11099.ml -fold 1 -fstout $GUTTDIR/CguttatusFstMB1910
        -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1910.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1910.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1910.saf.pos.gz
        -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1927.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1927.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1927.saf.pos.gz
        -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB27B.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB27B.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB27B.saf.pos.gz
        -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/zo11099.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/zo11099.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/zo11099.saf.pos.gz
        -> args: tole:0.000001 nthreads:4 maxiter:100 nsites:0 start:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1910.MB1927.ml chr:(null) start:-1 stop:-1 fstout:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/CguttatusFstMB1910 oldout:0 seed:1594319535 bootstrap:0 resample_chr:0 whichFst:0 fold:1 ref:(null) anc:(null)
        -> nSites: 100000
        -> Never here
```

When this happens the output stops with "Never here" and no fstout is generated.

In this case I know that the issue must be related to either MB1910.saf.idx (or one of the .ml files) because I also have used this command (see below) with the other .saf.idx files, and it worked and produced the fstout file. The beginning of my output from this successful command is shown below: instead of ending with "Never here," it gave the warning about not using a folded saf and proceeded with the rest of the analysis (output not shown).
```
$ANGSDIR/misc/realSFS fst index $GUTTDIR/MB1927.saf.idx $GUTTDIR/MB27B.saf.idx $GUTTDIR/zo11099.saf.idx -sfs $GUTTDIR/MB1927.MB27B.ml -sfs $GUTTDIR/MB1927.zo11099.ml -sfs $GUTTDIR/MB27B.zo11099.ml -fold 1 -fstout $GUTTDIR/CguttatusFstMB1927                                                                                                                                                                    -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1927.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1927.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1927.saf.pos.gz
        -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB27B.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB27B.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB27B.saf.pos.gz
        -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/zo11099.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/zo11099.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/zo11099.saf.pos.gz
        -> args: tole:0.000001 nthreads:4 maxiter:100 nsites:0 start:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1927.MB27B.ml chr:(null) start:-1 stop:-1 fstout:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/CguttatusFstMB1927 oldout:0 seed:1594320611 bootstrap:0 resample_chr:0 whichFst:0 fold:1 ref:(null) anc:(null)
        -> nSites: 100000
        -> IMPORTANT: please make sure that your saf files hasnt been folded with -fold 1 in -doSaf in angsd
```

The .ml files from code that produces "Never here" look fine to me. I also did this:

```
$ANGSDIR/misc/realSFS print $GUTTDIR/MB1910.saf.idx
```
And it printed out a ton of information (so at least the file can be opened and seems to contain valid information).

Any tips for how to troubleshoot this one? 
Thanks!
-Teresa

**Comment by TeresaPegan** [comment]
Do I need to include a .ml file for _every_ possible binary combination of .saf.idx files?  (I had left some combinations out in my code above.)

In any case, I tried again with my .ml files for every combination and I still get "Never here."

```
$ANGSDIR/misc/realSFS fst index $GUTTDIR/MB1910.saf.idx $GUTTDIR/MB1927.saf.idx $GUTTDIR/MB27B.saf.idx $GUTTDIR/zo11099.saf.idx -sfs $GUTTDIR/MB1910.MB1927.ml -sfs $GUTTDIR/MB1910.MB27B.ml -sfs $GUTTDIR/MB1910.zo11099.ml -sfs $GUTTDIR/MB1927.MB27B.ml -sfs $GUTTDIR/MB1927.zo11099.ml -sfs $GUTTDIR/MB27B.zo11099.ml -fold 1 -fstout $GUTTDIR/CguttatusFstMB1910
        -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1910.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1910.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1910.saf.pos.gz
        -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1927.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1927.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1927.saf.pos.gz
        -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB27B.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB27B.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB27B.saf.pos.gz
        -> Version of fname:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/zo11099.saf.idx is:2
        -> Assuming .saf.gz file: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/zo11099.saf.gz
        -> Assuming .saf.pos.gz: WGLC_Plate_1/C_guttatus_aligned_zebrafinch/zo11099.saf.pos.gz
        -> args: tole:0.000001 nthreads:4 maxiter:100 nsites:0 start:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/MB1910.MB1927.ml chr:(null) start:-1 stop:-1 fstout:WGLC_Plate_1/C_guttatus_aligned_zebrafinch/CguttatusFstMB1910 oldout:0 seed:1594390215 bootstrap:0 resample_chr:0 whichFst:0 fold:1 ref:(null) anc:(null)
        -> nSites: 100000
        -> Never here
```
Thanks,
-Teresa

**Comment by mwhj** [comment]
I don't have a solution but want to raise the same issue.  My .ml files are fine and I've used them to successfully follow the pairwise examples with my own data:

`export PATH=/data/genomicsocorg/mwhj1/Programs/ANGSD_Jun_2020/angsd:$PATH      
export PATH=/data/genomicsocorg/mwhj1/Programs/ANGSD_Jun_2020/angsd/misc:$PATH  

datadir=/scratch/genomicsocorg/mwhj1/ANGSD_OTref_Fst/Data
cd $datadir

p1=
p2=

echo Working with populations
echo $p1
echo and
echo $p2

p1p2="$p1$p2"

echo Calculating 2dSFS prior
saf1=${p1}_saf_folded.saf.idx
saf2=${p2}_saf_folded.saf.idx
/data/genomicsocorg/mwhj1/Programs/ANGSD_Jun_2020/angsd/misc/realSFS \
$saf1 $saf2 -P 16 -fold 1 > ${p1p2}.ml && echo Finished 2dSFS prior

echo Preparing fst for easy window analysis
/data/genomicsocorg/mwhj1/Programs/ANGSD_Jun_2020/angsd/misc/realSFS \
fst index $saf1 $saf2 -sfs ${p1p2}.ml -fstout $p1p2 && echo Finished preparing for window analysis

echo Getting global Fst estimate
/data/genomicsocorg/mwhj1/Programs/ANGSD_Jun_2020/angsd/misc/realSFS \
fst stats ${p1p2}.fst.idx > ${p1p2}.fst.idx.stats && echo Finished global Fst estimate

echo Starting sliding window analysis
/data/genomicsocorg/mwhj1/Programs/ANGSD_Jun_2020/angsd/misc/realSFS \
fst stats2 ${p1p2}.fst.idx -win 50000 -step 10000 > ${p1p2}.fst.idx.slidingwindow && echo Finished sliding window analysis && echo ...Done`

This worked for 5 populations of data without any problems.

This issue arises only when running realsfs index on n>2 populations (I use the .ml files created in the first step above).  

`export PATH=/data/genomicsocorg/mwhj1/Programs/ANGSD_and_dependencies/angsd:$PATH

export PATH=/data/genomicsocorg/mwhj1/Programs/ANGSD_and_dependencies/angsd/misc:$PATH

datadir=/scratch/genomicsocorg/mwhj1/ANGSD_OTref_Fst/Data
cd $datadir

echo Preparing fst for easy window analysis
/data/genomicsocorg/mwhj1/Programs/ANGSD_Jun_2020/angsd/misc/realSFS \
fst index L_saf_folded.saf.idx NR_saf_folded.saf.idx OT_saf_folded.saf.idx \
PF_saf_folded.saf.idx V_saf_folded.saf.idx \
-sfs LNR.ml -sfs LOT.ml -sfs LPF.ml -sfs LV.ml -sfs NROT.ml -sfs NRPF.ml \
-sfs NRV.ml -sfs OTPF.ml -sfs OTV.ml -sfs PFV.ml \
-fold 1 -fstout all_vs_all && echo Finished preparing for window analysis

echo Getting global Fst estimate
/data/genomicsocorg/mwhj1/Programs/ANGSD_Jun_2020/angsd/misc/realSFS \
fst stats all_vs_all.fst.idx > all_vs_all.fst.idx.stats && echo Finished global Fst estimate

echo Starting sliding window analysis
/data/genomicsocorg/mwhj1/Programs/ANGSD_Jun_2020/angsd/misc/realSFS \
fst stats2 all_vs_all.fst.idx \
-win 1 -step 1 > all_vs_all.fst.idx.slidingwindow_1step && echo Finished sliding window analysis && echo ...Done`

....

`Preparing fst for easy window analysis
	-> Version of fname:L_saf_folded.saf.idx is:2
	-> Assuming .saf.gz file: L_saf_folded.saf.gz
	-> Assuming .saf.pos.gz: L_saf_folded.saf.pos.gz
	-> Version of fname:NR_saf_folded.saf.idx is:2
	-> Assuming .saf.gz file: NR_saf_folded.saf.gz
	-> Assuming .saf.pos.gz: NR_saf_folded.saf.pos.gz
	-> Version of fname:OT_saf_folded.saf.idx is:2
	-> Assuming .saf.gz file: OT_saf_folded.saf.gz
	-> Assuming .saf.pos.gz: OT_saf_folded.saf.pos.gz
	-> Version of fname:PF_saf_folded.saf.idx is:2
	-> Assuming .saf.gz file: PF_saf_folded.saf.gz
	-> Assuming .saf.pos.gz: PF_saf_folded.saf.pos.gz
	-> Version of fname:V_saf_folded.saf.idx is:2
	-> Assuming .saf.gz file: V_saf_folded.saf.gz
	-> Assuming .saf.pos.gz: V_saf_folded.saf.pos.gz
	-> args: tole:0.000001 nthreads:4 maxiter:100 nsites:0 start:LNR.ml chr:(null) start:-1 stop:-1 fstout:all_vs_all oldout:0 seed:1595327900 bootstrap:0 resample_chr:0 whichFst:0 fold:1 ref:(null) anc:(null)
	-> nSites: 100000
	-> Never here
Finished preparing for window analysis
Getting global Fst estimate
	-> Assuming idxname:all_vs_all.fst.idx
	-> Problem opening file: 'all_vs_all.fst.idx'
Finished global Fst estimate
Starting sliding window analysis
	-> Assuming idxname:all_vs_all.fst.idx
	-> Problem opening file: 'all_vs_all.fst.idx'
Finished sliding window analysis
...Done`

I am running:

`angsd version: 0.930 (htslib: 1.9) build(Jun 29 2020 12:43:01)`

Any help at all would be very deeply appreciated.

**Comment by TeresaPegan** [comment]
I have gotten around this issue by just calculated each pairwise comparison separately - in both your example and mine, the problem arises when you try to do pairwise comparisons for more than 3 populations at once, I think? 
In my case this involves submitting an array job on my high-performance computing cluster that makes each pairwise calculation in parallel. The script I submitted calculates a pairwise comparison for just two of my populations for each job in the array. I made a text file with every possible population pair and submitted as many jobs as there were lines in that file (each array job took a line from the text file and used it to determine which populations to use). Then it's pretty simple to concatenate each pair's output files into one file. Hopefully that is helpful to know, let me know if you have any questions.

**Comment by mwhj** [comment]
Ah that is interesting!  Thank you for replying to me so quickly!  

I'm not certain, but it sounds like I've done that already for all of my population pairs:

`realSFS fst index $saf1 $saf2 -sfs ${p1p2}.ml -fstout $p1p2`

Is that the kind of thing you mean?  If so, is it really just a case of running something like

`cat *.ml >> all.ml` ?

**Comment by mwhj** [comment]
Ah no I can see that's a really stupid question, sorry - if you could walk me through it I would be incredibly grateful.

**Comment by TeresaPegan** [comment]
Actually I did the whole process in parallel, from realSFS fst index onwards, and concatenated the Fst output. My goal was just to get 1 measure of pairwise Fst per population pair, and it works to do it this way. 
Each slurm array job does this:
Read in a population pair from $MAP, which is my text file of all possible pairs. If this were slurm array job # 1, the task ID would be 1 and it would read line 1 of that text file. My populations are assigned as IND1 and IND2 in these scripts.
```
IND1=$(awk -F"," -v n=$SLURM_ARRAY_TASK_ID 'NR == n {print $1}' $MAP)
IND2=$(awk -F"," -v n=$SLURM_ARRAY_TASK_ID 'NR == n {print $2}' $MAP)
```
$MAP is just a simple file like this:

> Pop1   Pop2
> Pop1   Pop3
> Pop2   Pop3

etc.
Next, do the SFS step with SAF files that I already created in a previous script. These are named with the same population names as were in $MAP, so I can use my $IND variables to select them. Also, I am folding my SFS. $ANSGDIR and $GUTTDIR are just path shortcuts.
```
$ANGSDIR/misc/realSFS -fold 1 $GUTTDIR/$IND1".saf.idx" $GUTTDIR/$IND2".saf.idx" -P 4 > $GUTTDIR/$IND1"."$IND2".ml"
```
Fst index step with the existing SAF files and the ml file just created:
```
$ANGSDIR/misc/realSFS fst index -fold 1 $GUTTDIR/$IND1".saf.idx" $GUTTDIR/$IND2".saf.idx" -sfs $GUTTDIR/$IND1"."$IND2".ml" -fstout $GUTTDIR/$IND1"."$IND2
```
Calculate Fst for this pair:
```
$ANGSDIR/misc/realSFS fst stats $GUTTDIR/$IND1"."$IND2".fst.idx" > $GUTTDIR/$IND1"."$IND2".out"
```

Later, take all .out files from my analysis folder, each of which has Fst for one pair, and combine them together with a column for filename (which in my script include the identity of the two populations in each pair).
```
awk '{print $0,FILENAME}' *.out > pairwise_Fst.txt
```

Hope that helps! 

p.s. regarding the formatting, I think you want three "`" marks at the top and bottom of a code block, not one. If you click "insert code" it automatically puts one, but I think with only one you don't get to retain the formatting of the code, whereas with three all of the line breaks and indentations are kept.
